recipes/openai.py

```diff
@@ -0,0 +1,378 @@
+import abc
+import copy
+import os
+import sys
+import time
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Callable, Dict, Iterable, List, Optional, Tuple, TypeVar
+
+import httpx
+import jinja2
+import prodigy
+import rich
+import srsly
+from prodigy.components import preprocess
+from prodigy.util import msg, set_hashes
+from rich.panel import Panel
+from spacy.language import Language
+
+_ItemT = TypeVar("_ItemT")
+
+
+@dataclass
+class PromptExample(abc.ABC):
+    """An example to be passed into an OpenAI prompt."""
+
+    @staticmethod
+    def is_flagged(example: Dict) -> bool:
+        """Check whether a Prodigy example is flagged for use
+        in the prompt."""
+
+        return (
+            example.get("flagged") is True
+            and example.get("answer") == "accept"
+            and "text" in example
+        )
+
+    @abc.abstractclassmethod
+    def from_prodigy(cls, example: Dict, labels: Iterable[str]) -> "PromptExample":
+        """Create a prompt example from Prodigy's format."""
+        pass
+
+
+def normalize_label(label: str) -> str:
+    return label.lower()
+
+
+class OpenAISuggester:
+    """Suggest annotations using OpenAI's ChatGPT
+
+    prompt_template (jinja2.Template): A Jinja2 template that contains the
+        prompt to send to OpenAI's ChatGPT API.
+    model (str): The ChatGPT model ID to use for completion. Check the OpenAI
+        documentation for more information https://beta.openai.com/docs/models/overview.
+    labels (List[str]): List of labels for annotation.
+    max_examples (int): The maximum number of examples to stream in the Prodigy UI.
+    segment (bool): If set to True, segment the documents into sentences.
+    verbose (bool): Show verbose output in the command-line, including the prompt and response from OpenAI.
+    openai_api_org (str): The OpenAI API organization.
+    openai_api_key (str): The OpenAI API key.
+    openai_temperature (float): The temperature parameter (from 0 to 1) that controls the
+        randomness of ChatGPT's output.
+    openai_max_tokens (int): The maximum amout of tokens that ChatGPT's
+        completion API can generate.
+    openai_n (int): The number of completions to generate for each prompt.
+    openai_n_retries (int): The number of retries whenever a 429 error occurs.
+    openai_retry_timeout_s (int): The amount of time before attempting another request whenever we
+        encounter a 429 error. Increases exponentially for each retry.
+    openai_read_timeout_s (int): The amount of time to wait a response output during a request.
+    examples (List[PromptExample]): A list of examples to add to the prompt to guide ChatGPT output.
+    response_parser (Callable[str] -> Dict): A function that accepts a string that represents
+        ChatGPT's raw response, and parses it into a dictionary that is compatible to Prodigy's
+        annotation interfaces.
```

Comment:
Added some docstrings. Feel free to correct or suggest changes whenever needed 👍